File: solved/typescript/lowestCommonAncestor/lowestCommonAncestor.ts
// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/

export class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor (val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val
    this.left = left === undefined ? null : left
    this.right = right === undefined ? null : right
  }
}

export const lowestCommonAncestor = (
  root: TreeNode | null,
  p: TreeNode | null,
  q: TreeNode | null
): TreeNode => {
  return (root!.val - p!.val) * (root!.val - q!.val) < 1
    ? root!
    : lowestCommonAncestor(
        p!.val < root!.val ? root!.left! : root!.right!,
        p!,
        q!
      )
}

File: solved/LowestCommonAncestor.go
// Package main has solved algorithms
// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/solution/
package main

// import log "github.com/sirupsen/logrus"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil || root == p || root == q {
		return root
	}

	left := lowestCommonAncestor(root.Left, p, q)
	right := lowestCommonAncestor(root.Right, p, q)
	if left != nil && right != nil {
		return root
	}
	if left != nil {
		return left
	}
	return right
}

func main() {
	// log.Info(lowestCommonAncestor())
}
